Add self-assignment and multiple-parent checks to Resource.assign_child_resource

naikymen opened the pull request:
Motivation: resources can be assigned to themseves and have multiple parents.

![image](https://user-images.githubusercontent.com/3259326/229376237-ad9909f3-bc94-4c19-8e42-85df93576e2b.png)

Which can be fixed by this PR.

rickwierenga commented on pylabrobot/resources/resource.py:
```diff
@@ -136,16 +136,34 @@ def assign_child_resource(self, resource: Resource, location: Optional[Coordinat
+      msgs.append(f"Will not assign resource '{self.name}' to itself.")
+
+    # Check for multiple parents
+    if resource.parent:
```
`if resource.parent is not None` is cleaner.

naikymen replied:
Yes :)

rickwierenga commented on pylabrobot/resources/resource.py:
```diff
@@ -136,16 +136,34 @@ def assign_child_resource(self, resource: Resource, location: Optional[Coordinat
     self.children.append(resource)
+    
+  def check_assignment(self, resource: Resource, location: Optional[Coordinate]):
+    msgs = []
```
These errors are mutually exclusive (one would be raised before the other, so no two conditions should be true at the same time), so it may be better to raise the errors directly rather than putting them in a list.

naikymen replied:
Hmm both can be true if the resource has a parent, and the user is trying to assign it it itself.

I made a list just to get every "wrong thing" into the message, thinking that more checks might be added in the future.

rickwierenga replied:
Right, let's keep the list for now.

rickwierenga commented on pylabrobot/resources/resource.py:
```diff
@@ -136,16 +136,34 @@ def assign_child_resource(self, resource: Resource, location: Optional[Coordinat
-    resource.parent = self
-    resource.location = location
     try:
+      self.check_assignment(resource, location)
```
`check_assignment` is sufficiently short so that we can put it above this try block. Happy to discuss.

naikymen replied:
Sure, I only wanted it above this, to catch the errors before modifying the objects.

```
      resource.parent = self
      resource.location = location
```

I put it into the try clause just to reuse it, not a strong preference.

Im not confident about the place. Perhaps you can choose a better one?

Perhaps inside the callback? Which seems to do a name check already.

rickwierenga replied:
I think it's fine to put it above the assignment of `location` and `parent`, as the first code in this method. That way we do not need set the values to `None` if an error occurs.

The reason I assigned them before calling the callback (and unassign them on error), is that some callbacks will want to use this information do determine validity. That is not the case here.

naikymen replied:
Done :ok:

rickwierenga reviewed (changes requested):
Thanks for the PR! Good finds, these error messages definitely make things clearer for the user.

I just had a few small comments about the code.

naikymen commented:
That was a quick response :)

I pushed an additional fix for the second proposed check.

rickwierenga commented:
Unit tests are failing because `assign_child_resource` used to override an existing parent. I think it's good to keep that behavior, but to make it optional using an `reassign: bool` parameter. Based on this parameter, an error would be raised if the resource already has a parent. What do you think is best: defaulting to true or false?

naikymen commented:
> Unit tests are failing because `assign_child_resource` used to override an existing parent. I think it's good to keep that behavior, but to make it optional using an `reassign: bool` parameter. Based on this parameter, an error would be raised if the resource already has a parent. What do you think is best: defaulting to true or false?

A re-assign behaviour with an existing parent makes sense to me too.

I'm not sure I understand completely though. Where should the reassign parameter be added?

rickwierenga commented:
To the `assign_child_resource` method, with a default value (true or false) so the user is not required to specify it. I'm not sure what default value I like best.

naikymen commented:
I see, hmm.

I added the check because I wanted to know if an instance of a resource could be lazily reused to add multiple similar resources to a deck.

That was why I would expect to see an error in my setup, people tend to be lazy.

On the other hand, I'm also not expecting stuff to move labware around (which might happen in the future). In the movable labware case the user will likely know that they're doing by reassignment.

Adding parameters and new functions complicates things. I think that `reassign=True` as default is good and simple to understand.

Perhaps a better check would be against re-assignment to the same parent, which would align with my original motivation, but allow reassignments.

naikymen commented:
However, that last case is already covered by a downstream exception "Error: Resource 'plate_01' already assigned to deck"

The only "trouble" with that implementation is that the resource's parent and location are reset to None. This is the only point in favour of writing a check for this case before the try clause.

naikymen commented:
Added the `reassign` parameter to `assign_child_resource` with default `True` (replicates the default behaviour). Stuff should not break.

Added a check on reassigning to the same parent, which is harmless by may be due to misuse of PLR. Not sure if this can break other things.

rickwierenga commented on pylabrobot/resources/resource.py:
```diff
@@ -147,6 +153,31 @@ def assign_child_resource(self, resource: Resource, location: Optional[Coordinat
+
+    # Check for reassignment to the same parent
+    if resource.parent is not None and (resource.parent is self):
+      msgs.append(f"Will not reassign resource '{resource.name}' " +
+                  f"to the same parent: '{resource.parent.name}'.")
```
Perhaps it's better to make this a warning (using [the logger](https://github.com/PyLabRobot/pylabrobot/blob/86ff89651bc41de093a9840613ba7a78136b55fe/pylabrobot/__init__.py#L7))? Something like `f"Resource '{resource.name}' already assigned to '{resource.parent.name}'"`. I'm somewhat hesitant to raising errors when something is technically OK.

It would crash if `reassign` is False.

naikymen replied:
Done :)

Note that now the downstream check (in `Deck._check_name_exists`)  is raised instead.

![image](https://user-images.githubusercontent.com/3259326/229383489-b3ccd25b-69bd-4447-9bef-6c5325d0c482.png)

The logger shows: 

`2023-04-02 19:53:24,904 - pylabrobot.resources.resource - WARNING - Resource 'plate_01' already assigned to 'basic deck 1'`

naikymen replied:
Output with reassign=False:

![image](https://user-images.githubusercontent.com/3259326/229383581-a2608257-3ced-44b2-813e-e5ff179ab1b0.png)

rickwierenga replied:
Great!

rickwierenga commented:
> On the other hand, I'm also not expecting stuff to move labware around (which might happen in the future). In the movable labware case the user will likely know that they're doing by reassignment.

Yes. It's probably just when defining deck layouts. Beyond that, I think this should be PLR's concern, for regular users at least.

> The only "trouble" with that implementation is that the resource's parent and location are reset to None. This is the only point in favour of writing a check for this case before the try clause.

Right. I have briefly considered adding two callbacks, perhaps `will_assign_resource_callback` and `did_assign_resource_callback`. I think this is a good motivation for finally implementing that.

> Added the reassign parameter to assign_child_resource with default True (replicates the default behaviour). Other stuff should not break.

Nice!

> Added a check on reassigning to the same parent, which is harmless by may be due to misuse of PLR. Not sure if this can break other things.

Added a comment in the code.

rickwierenga commented:
Thanks again!
